apps/src/code-studio/stageLockRedux.js

```diff
@@ -0,0 +1,240 @@
+    const lockDialogStageId = action.stageId;
+
+    const students = sections[selectedSection].stages[lockDialogStageId];
+    return Object.assign({}, state, {
+      lockDialogStageId,
+      lockStatus: lockStatusForStage(state, action.stageId)
+    });
+  }
+
+  if (action.type === CLOSE_LOCK_DIALOG) {
+    return Object.assign({}, state, {
+      lockDialogStageId: null,
+      lockStatus: [],
+      saving: false
+    });
+  }
+
+  if (action.type === BEGIN_SAVE) {
+    return Object.assign({}, state, {
+      saving: true
+    });
+  }
+
+  if (action.type === FINISH_SAVE) {
+    const { sections, selectedSection } = state;
+    const { lockStatus: nextLockStatus, stageId } = action;
+    const nextStage = _.cloneDeep(sections[selectedSection].stages[stageId]);
+
+    // Update locked/view_answers in stages based on the new lockStatus provided
+    // by our dialog.
+    nextStage.forEach((item, index) => {
+      const update = nextLockStatus[index];
+      // We assume lockStatus is ordered the same as stageToUpdate. Let's
+      // validate that.
+      if (item.user_level_id !== update.userLevelId) {
+        throw new Error('Expect user ids be the same');
+      }
+      item.locked = update.lockStatus === LockStatus.Locked;
+      item.view_answers = update.lockStatus === LockStatus.Readonly;
+    });
+
+    const nextState = _.cloneDeep(state);
+    nextState.sections[selectedSection].stages[stageId] = nextStage;
+    nextState.lockStatus = nextLockStatus;
+    nextState.unlockedStageIds = unlockedStages(nextState.sections[selectedSection]);
+    return nextState;
+  }
+
+  return state;
+}
+
+// Action creators
+export const setViewType = viewType => {
+  if (!ViewType[viewType]) {
+    throw new Error('unknown ViewType: ' + viewType);
+  }
+
+  return {
+    type: SET_VIEW_TYPE,
+    viewAs: viewType
+  };
+};
+
+export const setSections = sections => ({
+  type: SET_SECTIONS,
+  sections
+});
+
+export const selectSection = sectionId => ({
+  type: SELECT_SECTION,
+  sectionId
+});
+
+export const openLockDialog = stageId => ({
+  type: OPEN_LOCK_DIALOG,
+  stageId
+});
+
+const beginSave = () => ({ type: BEGIN_SAVE });
+const finishSave = (newLockStatus, stageId) => ({
+  type: FINISH_SAVE,
+  lockStatus: newLockStatus,
+  stageId
+});
+
+/**
+ * Action asynchronously dispatches a set of actions around saving our
+ * lock status.
+ */
+const performSave = (newLockStatus, stageId) => {
+  return (dispatch, getState) => {
+    const oldLockStatus = getState().stageLock.lockStatus;
+
+    const saveData = newLockStatus.filter((item, index) => {
+      // Only need to save items that changed
+      return !_.isEqual(item, oldLockStatus[index]);
+    }).map(item => ({
+      user_level_id: item.userLevelId,
+      locked: item.lockStatus === LockStatus.Locked,
+      view_answers: item.lockStatus === LockStatus.Readonly
+    }));
+
+    if (saveData.length === 0) {
+      dispatch(closeLockDialog());
```

Comment:
I was surprised to find that this action also closes the dialog.  It might be cleaner for this action creator just to take an `onComplete` callback, and let the caller specify that the dialog should be closed when the save is done.